More details in snapshot debug dump

ChillerDragon opened the pull request:
The method CSnapShot::DebugDump() used to print just the raw snap items. Without item names or field names:

```
2024-04-07 09:10:01 I snapshot: data_size=356 num_items=10
2024-04-07 09:10:01 I snapshot:         type=0 id=32767
2024-04-07 09:10:01 I snapshot:                   0  -1824658838        933dea6a
2024-04-07 09:10:01 I snapshot:                   1   -629591830        da7930ea
2024-04-07 09:10:01 I snapshot:                   2  -1450210576        a98f8af0
2024-04-07 09:10:01 I snapshot:                   3    914991429        3689a945
2024-04-07 09:10:01 I snapshot:         type=0 id=32766
2024-04-07 09:10:01 I snapshot:                   0    583701389        22ca938d
2024-04-07 09:10:01 I snapshot:                   1    327171627        13803e2b
2024-04-07 09:10:01 I snapshot:                   2  -1636052395        9e7bd255
2024-04-07 09:10:01 I snapshot:                   3  -1901674991        8ea6be11
2024-04-07 09:10:01 I snapshot:         type=0 id=32765
2024-04-07 09:10:01 I snapshot:                   0   1993229659        76ce455b
2024-04-07 09:10:01 I snapshot:                   1   -102024632        f9eb3a48
2024-04-07 09:10:01 I snapshot:                   2  -1378361269        add7e04b
2024-04-07 09:10:01 I snapshot:                   3  -1810037668        941d045c
2024-04-07 09:10:01 I snapshot:         type=6 id=0
2024-04-07 09:10:01 I snapshot:                   0            0        00000000
2024-04-07 09:10:01 I snapshot:                   1            0        00000000
2024-04-07 09:10:01 I snapshot:                   2            0        00000000
2024-04-07 09:10:01 I snapshot:                   3            0        00000000
2024-04-07 09:10:01 I snapshot:                   4            0        00000000
2024-04-07 09:10:01 I snapshot:                   5            0        00000000
2024-04-07 09:10:01 I snapshot:                   6            0        00000000
2024-04-07 09:10:01 I snapshot:                   7            1        00000001
2024-04-07 09:10:01 I snapshot:         type=32767 id=0
2024-04-07 09:10:01 I snapshot:                   0    247983155        0ec7ec33
2024-04-07 09:10:01 I snapshot:                   1            9        00000009
2024-04-07 09:10:01 I snapshot:                   2           64        00000040
2024-04-07 09:10:01 I snapshot:         type=11 id=0
2024-04-07 09:10:01 I snapshot:                   0  -1008145940        c3e8e9ec
2024-04-07 09:10:01 I snapshot:                   1   -320474428        ece5f2c4
2024-04-07 09:10:01 I snapshot:                   2   -220076049        f2e1e7ef
2024-04-07 09:10:01 I snapshot:                   3   -293568512        ee808000
2024-04-07 09:10:01 I snapshot:                   4  -1330727040        b0aeb780
2024-04-07 09:10:01 I snapshot:                   5  -2139062144        80808080
2024-04-07 09:10:01 I snapshot:                   6  -2139062272        80808000
2024-04-07 09:10:01 I snapshot:                   7           64        00000040
2024-04-07 09:10:01 I snapshot:                   8   -403511835        e7f2e5e5
2024-04-07 09:10:01 I snapshot:                   9   -286001183        eef3f7e1
2024-04-07 09:10:01 I snapshot:                  10   -219905920        f2e48080
2024-04-07 09:10:01 I snapshot:                  11  -2139062144        80808080
2024-04-07 09:10:01 I snapshot:                  12  -2139062144        80808080
2024-04-07 09:10:01 I snapshot:                  13  -2139062272        80808000
2024-04-07 09:10:01 I snapshot:                  14            0        00000000
2024-04-07 09:10:01 I snapshot:                  15      7199301        006dda45
2024-04-07 09:10:01 I snapshot:                  16      2345678        0023cace
2024-04-07 09:10:01 I snapshot:         type=10 id=0
2024-04-07 09:10:01 I snapshot:                   0            1        00000001
2024-04-07 09:10:01 I snapshot:                   1            0        00000000
2024-04-07 09:10:01 I snapshot:                   2            0        00000000
2024-04-07 09:10:01 I snapshot:                   3        -9999        ffffd8f1
2024-04-07 09:10:01 I snapshot:                   4            0        00000000
2024-04-07 09:10:01 I snapshot:         type=32766 id=0
2024-04-07 09:10:01 I snapshot:                   0            1        00000001
2024-04-07 09:10:01 I snapshot:                   1            0        00000000
2024-04-07 09:10:01 I snapshot:         type=9 id=0
2024-04-07 09:10:01 I snapshot:                   0          438        000001b6
2024-04-07 09:10:01 I snapshot:                   1         1968        000007b0
2024-04-07 09:10:01 I snapshot:                   2          913        00000391
2024-04-07 09:10:01 I snapshot:                   3            0        00000000
2024-04-07 09:10:01 I snapshot:                   4            0        00000000
2024-04-07 09:10:01 I snapshot:                   5            0        00000000
2024-04-07 09:10:01 I snapshot:                   6            0        00000000
2024-04-07 09:10:01 I snapshot:                   7            0        00000000
2024-04-07 09:10:01 I snapshot:                   8           -1        ffffffff
2024-04-07 09:10:01 I snapshot:                   9            0        00000000
2024-04-07 09:10:01 I snapshot:                  10            0        00000000
2024-04-07 09:10:01 I snapshot:                  11         1968        000007b0
2024-04-07 09:10:01 I snapshot:                  12          913        00000391
2024-04-07 09:10:01 I snapshot:                  13            0        00000000
2024-04-07 09:10:01 I snapshot:                  14            0        00000000
2024-04-07 09:10:01 I snapshot:                  15            1        00000001
2024-04-07 09:10:01 I snapshot:                  16           10        0000000a
2024-04-07 09:10:01 I snapshot:                  17           10        0000000a
2024-04-07 09:10:01 I snapshot:                  18            0        00000000
2024-04-07 09:10:01 I snapshot:                  19            1        00000001
2024-04-07 09:10:01 I snapshot:                  20            5        00000005
2024-04-07 09:10:01 I snapshot:                  21            0        00000000
2024-04-07 09:10:01 I snapshot:         type=32765 id=0
2024-04-07 09:10:01 I snapshot:                   0        49152        0000c000
2024-04-07 09:10:01 I snapshot:                   1            0        00000000
2024-04-07 09:10:01 I snapshot:                   2            2        00000002
2024-04-07 09:10:01 I snapshot:                   3            0        00000000
2024-04-07 09:10:01 I snapshot:                   4            0        00000000
2024-04-07 09:10:01 I snapshot:                   5            0        00000000
2024-04-07 09:10:01 I snapshot:                   6            0        00000000
2024-04-07 09:10:01 I snapshot:                   7            0        00000000
2024-04-07 09:10:01 I snapshot:                   8            1        00000001
2024-04-07 09:10:01 I snapshot:                   9            0        00000000
```

Now it looks like this:

```
2024-04-07 09:21:42 I snapshot: data_size=356 num_items=10
2024-04-07 09:21:42 I snapshot:         EX/UUID type=0 id=32767 size=16
2024-04-07 09:21:42 I snapshot:                   0  -1824658838        933dea6a
2024-04-07 09:21:42 I snapshot:                   1   -629591830        da7930ea
2024-04-07 09:21:42 I snapshot:                   2  -1450210576        a98f8af0
2024-04-07 09:21:42 I snapshot:                   3    914991429        3689a945
2024-04-07 09:21:42 I snapshot:         EX/UUID type=0 id=32766 size=16
2024-04-07 09:21:42 I snapshot:                   0    583701389        22ca938d
2024-04-07 09:21:42 I snapshot:                   1    327171627        13803e2b
2024-04-07 09:21:42 I snapshot:                   2  -1636052395        9e7bd255
2024-04-07 09:21:42 I snapshot:                   3  -1901674991        8ea6be11
2024-04-07 09:21:42 I snapshot:         EX/UUID type=0 id=32765 size=16
2024-04-07 09:21:42 I snapshot:                   0   1993229659        76ce455b
2024-04-07 09:21:42 I snapshot:                   1   -102024632        f9eb3a48
2024-04-07 09:21:42 I snapshot:                   2  -1378361269        add7e04b
2024-04-07 09:21:42 I snapshot:                   3  -1810037668        941d045c
2024-04-07 09:21:42 I snapshot:         GameInfo type=6 id=0 size=32
2024-04-07 09:21:42 I snapshot:                   0            0        00000000        m_GameFlags=0 (min=0 max=256)
2024-04-07 09:21:42 I snapshot:                   1            0        00000000        m_GameStateFlags=0 (min=0 max=256)
2024-04-07 09:21:42 I snapshot:                   2            0        00000000        m_RoundStartTick=0 (NetTick)
2024-04-07 09:21:42 I snapshot:                   3            0        00000000        m_WarmupTimer=0 (min=min_int(-2147483648) max=max_int(2147483647))
2024-04-07 09:21:42 I snapshot:                   4            0        00000000        m_ScoreLimit=0 (min=0 max=max_int(2147483647))
2024-04-07 09:21:42 I snapshot:                   5            0        00000000        m_TimeLimit=0 (min=0 max=max_int(2147483647))
2024-04-07 09:21:42 I snapshot:                   6            0        00000000        m_RoundNum=0 (min=0 max=max_int(2147483647))
2024-04-07 09:21:42 I snapshot:                   7            1        00000001        m_RoundCurrent=1 (min=0 max=max_int(2147483647))
2024-04-07 09:21:42 I snapshot:         (out of range) type=32767 id=0 size=12
2024-04-07 09:21:42 I snapshot:                   0    247983155        0ec7ec33
2024-04-07 09:21:42 I snapshot:                   1            9        00000009
2024-04-07 09:21:42 I snapshot:                   2           64        00000040
2024-04-07 09:21:42 I snapshot:         ClientInfo type=11 id=0 size=68
2024-04-07 09:21:42 I snapshot:                   0  -1008145940        c3e8e9ec        m_Name0=-1008145940     IntsToStr: Chil
2024-04-07 09:21:42 I snapshot:                   1   -320474428        ece5f2c4        m_Name1=-320474428      IntsToStr: lerD
2024-04-07 09:21:42 I snapshot:                   2   -220076049        f2e1e7ef        m_Name2=-220076049      IntsToStr: rago
2024-04-07 09:21:42 I snapshot:                   3   -293568512        ee808000        m_Name3=-293568512      IntsToStr: n
2024-04-07 09:21:42 I snapshot:                   4  -1330727040        b0aeb780        m_Clan0=-1330727040     IntsToStr: 0.7
2024-04-07 09:21:42 I snapshot:                   5  -2139062144        80808080        m_Clan1=-2139062144     IntsToStr:
2024-04-07 09:21:42 I snapshot:                   6  -2139062272        80808000        m_Clan2=-2139062272     IntsToStr:
2024-04-07 09:21:42 I snapshot:                   7           64        00000040        m_Country=64
2024-04-07 09:21:42 I snapshot:                   8   -403511835        e7f2e5e5        m_Skin0=-403511835      IntsToStr: gree
2024-04-07 09:21:42 I snapshot:                   9   -286001183        eef3f7e1        m_Skin1=-286001183      IntsToStr: nswa
2024-04-07 09:21:42 I snapshot:                  10   -219905920        f2e48080        m_Skin2=-219905920      IntsToStr: rd
2024-04-07 09:21:42 I snapshot:                  11  -2139062144        80808080        m_Skin3=-2139062144     IntsToStr:
2024-04-07 09:21:42 I snapshot:                  12  -2139062144        80808080        m_Skin4=-2139062144     IntsToStr:
2024-04-07 09:21:42 I snapshot:                  13  -2139062272        80808000        m_Skin5=-2139062272     IntsToStr:
2024-04-07 09:21:42 I snapshot:                  14            0        00000000        m_UseCustomColor=0 (min=0 max=1)
2024-04-07 09:21:42 I snapshot:                  15      7199301        006dda45        m_ColorBody=7199301
2024-04-07 09:21:42 I snapshot:                  16      2345678        0023cace        m_ColorFeet=2345678
2024-04-07 09:21:42 I snapshot:         PlayerInfo type=10 id=0 size=20
2024-04-07 09:21:42 I snapshot:                   0            1        00000001        m_Local=1 (min=0 max=1)
2024-04-07 09:21:42 I snapshot:                   1            0        00000000        m_ClientId=0 (min=0 max=MAX_CLIENTS-1(63))
2024-04-07 09:21:42 I snapshot:                   2            0        00000000        m_Team=0 (min=TEAM_SPECTATORS(-1) max=TEAM_BLUE(1))
2024-04-07 09:21:42 I snapshot:                   3        -9999        ffffd8f1        m_Score=-9999
2024-04-07 09:21:42 I snapshot:                   4            0        00000000        m_Latency=0
2024-04-07 09:21:42 I snapshot:         (out of range) type=32766 id=0 size=8
2024-04-07 09:21:42 I snapshot:                   0            1        00000001
2024-04-07 09:21:42 I snapshot:                   1            0        00000000
2024-04-07 09:21:42 I snapshot:         Character type=9 id=0 size=88
2024-04-07 09:21:42 I snapshot:                   0          176        000000b0        m_Tick=176
2024-04-07 09:21:42 I snapshot:                   1         1968        000007b0        m_X=1968
2024-04-07 09:21:42 I snapshot:                   2          913        00000391        m_Y=913
2024-04-07 09:21:42 I snapshot:                   3            0        00000000        m_VelX=0
2024-04-07 09:21:42 I snapshot:                   4            0        00000000        m_VelY=0
2024-04-07 09:21:42 I snapshot:                   5            0        00000000        m_Angle=0
2024-04-07 09:21:42 I snapshot:                   6            0        00000000        m_Direction=0 (min=-1 max=1)
2024-04-07 09:21:42 I snapshot:                   7            0        00000000        m_Jumped=0 (min=0 max=3)
2024-04-07 09:21:42 I snapshot:                   8           -1        ffffffff        m_HookedPlayer=-1 (min=-1 max=MAX_CLIENTS-1(63))
2024-04-07 09:21:42 I snapshot:                   9            0        00000000        m_HookState=0 (min=-1 max=5)
2024-04-07 09:21:42 I snapshot:                  10            0        00000000        m_HookTick=0 (NetTick)
2024-04-07 09:21:42 I snapshot:                  11         1968        000007b0        m_HookX=1968
2024-04-07 09:21:42 I snapshot:                  12          913        00000391        m_HookY=913
2024-04-07 09:21:42 I snapshot:                  13            0        00000000        m_HookDx=0
2024-04-07 09:21:42 I snapshot:                  14            0        00000000        m_HookDy=0
2024-04-07 09:21:42 I snapshot:                  15            1        00000001        m_PlayerFlags=1 (min=0 max=256)
2024-04-07 09:21:42 I snapshot:                  16           10        0000000a        m_Health=10 (min=0 max=10)
2024-04-07 09:21:42 I snapshot:                  17           10        0000000a        m_Armor=10 (min=0 max=10)
2024-04-07 09:21:42 I snapshot:                  18            0        00000000        m_AmmoCount=0 (min=0 max=10)
2024-04-07 09:21:42 I snapshot:                  19            1        00000001        m_Weapon=1 (min=0 max=NUM_WEAPONS-1(5))
2024-04-07 09:21:42 I snapshot:                  20            5        00000005        m_Emote=5 (min=0 max=6)
2024-04-07 09:21:42 I snapshot:                  21            0        00000000        m_AttackTick=0 (min=0 max=max_int(2147483647))
2024-04-07 09:21:42 I snapshot:         (out of range) type=32765 id=0 size=40
2024-04-07 09:21:42 I snapshot:                   0        49152        0000c000
2024-04-07 09:21:42 I snapshot:                   1            0        00000000
2024-04-07 09:21:42 I snapshot:                   2            2        00000002
2024-04-07 09:21:42 I snapshot:                   3            0        00000000
2024-04-07 09:21:42 I snapshot:                   4            0        00000000
2024-04-07 09:21:42 I snapshot:                   5            0        00000000
2024-04-07 09:21:42 I snapshot:                   6            0        00000000
2024-04-07 09:21:42 I snapshot:                   7            0        00000000
2024-04-07 09:21:42 I snapshot:                   8            1        00000001
2024-04-07 09:21:42 I snapshot:                   9            0        00000000
```

https://github.com/ddnet/ddnet/assets/20344300/1f5cfa82-8956-4441-9341-f15405d4b5c0



https://github.com/ddnet/ddnet/assets/20344300/730e2089-15a3-4fca-9ef0-9dd526b9119e





## Checklist

- [ ] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

heinrich5991 commented on CMakeLists.txt:
```diff
@@ -1941,6 +1941,8 @@ set_src(ENGINE_SHARED GLOB_RECURSE src/engine/shared
   huffman.cpp
   huffman.h
+  int_strings.cpp
+  int_strings.h
```
Please name the files in a way that makes it clear that this is not a generic int-string conversion, but a tw-specific format that shouldn't be used.

ChillerDragon replied:
Suggestions?

ChillerDragon replied:
I went with **src/engine/shared/legacy_tw_int_strings.h** now :shrug:

heinrich5991 commented on src/engine/shared/int_strings.h:
```diff
@@ -0,0 +1,12 @@
+
+// Do not use these function unless for legacy code!
+void StrToInts(int *pInts, size_t NumInts, const char *pStr);
+bool IntsToStr(const int *pInts, size_t NumInts, char *pStr, size_t StrSize);
```
These functions should be renamed if you move them into engine, to make sure they're not accidentally used.

ChillerDragon replied:
I prefixed the function names with Legacy and also duplicated robstis legacy comment. So IDEs show the "Do not use" note for both functions.

heinrich5991 commented on src/engine/shared/int_strings.h:
```diff
@@ -0,0 +1,12 @@
+/* (c) Magnus Auvinen. See licence.txt in the root of the distribution for more information. */
+/* If you are missing that file, acquire a complete release at teeworlds.com.                */
```
This file is not copyrighted by Magnus Auvinen.

heinrich5991 commented on src/engine/shared/snapshot.h:
```diff
@@ -3,6 +3,9 @@
 #define ENGINE_SHARED_SNAPSHOT_H
 
+#include <game/generated/protocol.h>
+#include <game/generated/protocol7.h>
```
`engine` files must not depend on game headers.

heinrich5991 commented on src/engine/shared/snapshot.h:
```diff
@@ -49,6 +52,7 @@ class CSnapshot
 	};
 
 	int NumItems() const { return m_NumItems; }
+	int DataSize() const { return m_DataSize; }
```
That seems irrelevant to the PR.

ChillerDragon replied:
That is from a prior iteration where the code was somewhere in client.cpp not in snapshot. So it needed access. So this has to stay when you want me to move it out of the snapshot class again.

heinrich5991 commented on src/engine/shared/snapshot.h:
```diff
@@ -57,10 +61,13 @@ class CSnapshot
 
 	unsigned Crc() const;
-	void DebugDump() const;
+	void DebugDump(bool Sixup) const;
```
Inverting the control might be better. Implement a new function on `CNetObjHandler` that can debug-dump the snapshot. That way, engine stuff does not depend on game.

ChillerDragon replied:
done.

heinrich5991 commented on datasrc/datatypes.py:
```diff
@@ -383,16 +411,45 @@ def emit_unpack_msg(self):
+		for i in range(3):
+			str_vars.append(f"m_Clan{i}")
+		for i in range(6):
+			str_vars.append(f"m_Skin{i}")
```
Instead of weirdly fixing up the fields after the fact, change the type of the from separate `NetIntAny` variables to some `NetTwIntString` type or similar.

ChillerDragon replied:
Same as the filename. Can we come up with some cool name? Or should we go with TwIntString?

ChillerDragon replied:
Went with NetTwIntString :shrug:

heinrich5991 reviewed (commented):
The concept sounds very useful for debugging.

From the code, `engine` must not depend on `game` and matching fields by name doesn't sound good to me.

ChillerDragon commented on src/engine/shared/int_strings.cpp:
```diff
@@ -0,0 +1,54 @@
+		StrIndex++;
+	}
+	// Ensure null-termination
+	pStr[StrIndex] = '\0';
```
By the way non obvious in the git diff but I changed this line of code. It used to say ``pStr[StrIndex - 1] = '\0';`` but that cut off the int strings when only one integer was used.

ChillerDragon replied:
Okay I decided against messing with the int unpacker code :D it seemed like the only thing that broke was the test but I probably failed to test some edge cases in game.

Anyways I added a new function just for the snapshot annotation ``LegacyIntToStrPartial()`` which only takes one integer and prints the full string with all bytes. This can now be applied to every int in the snap individually and no letters get replaced by nullbytes anymore.

Robyt3 replied:
See https://github.com/ddnet/ddnet/pull/8189#discussion_r1556159092, `LegacyIntToStrPartial` should be unnecessary. You can unpack all 4 bytes of one int by adding another padding `int` with value `0x80808000`.

ChillerDragon commented:
I reverted the changes to CSnapshot::DebugDump() and added a new method to the generated code as @heinrich5991 suggested. The diff I used to debug print became a bit more complex but I think it is still fine to quickly add those lines for debugging something.

```diff
diff --git a/src/engine/client/client.cpp b/src/engine/client/client.cpp
index 919397300..49bc517cb 100644
--- a/src/engine/client/client.cpp
+++ b/src/engine/client/client.cpp
@@ -1815,6 +1815,8 @@ void CClient::ProcessServerPacket(CNetChunk *pPacket, int Conn, bool Dummy)
                                                return;
                                        }
 
+                                       GameClient()->GetNetObjHandler()->DebugDumpSnapshot(pTmpBuffer3);
+
                                        // add new
                                        m_aSnapshotStorage[Conn].Add(GameTick, time_get(), SnapSize, pTmpBuffer3, AltSnapSize, pAltSnapBuffer);
 
diff --git a/src/engine/server/server.cpp b/src/engine/server/server.cpp
index f258dea50..d8010d31a 100644
--- a/src/engine/server/server.cpp
+++ b/src/engine/server/server.cpp
@@ -928,6 +928,12 @@ void CServer::DoSnapshot()
                                m_aDemoRecorder[i].RecordSnapshot(Tick(), aData, SnapshotSize);
                        }
 
+                       if(m_aClients[i].m_Sixup)
+                               GameServer()->GetNetObjHandler7()->DebugDumpSnapshot(pData);
+                       else
+                               GameServer()->GetNetObjHandler()->DebugDumpSnapshot(pData);
+
+
                        int Crc = pData->Crc();
 
                        // remove old snapshots
```

The thing that bothers me more is that I expect the debug dump method for the snapshot to be in the snapshot class. Every time I was wondering about how snaps worked I found my self browsing snapshot.cpp and not the generated protocol.cpp. I left the old DebugDump() in place in case someone has no CNetObjHandler available. And added a big marketing comment to the function signature saying that there is also ``CNetObjHandler::DebugDumpSnapshot(const CSnapshot *pSnap)``

![image](https://github.com/ddnet/ddnet/assets/20344300/d9ace356-52bf-4315-adfb-362d1be0971a)

Robyt3 commented on src/engine/shared/legacy_tw_int_strings.cpp:
```diff
@@ -0,0 +1,81 @@
+	return false;
+}
+
+bool LegacyIntToStrPartial(int SrcInt, char *pDstStr, size_t DstStrSize)
+{
+	dbg_assert(DstStrSize - 1 >= sizeof(int), "LegacyIntToStrPartial: StrSize invalid");
+
+	// Unpack string without validation
+	size_t StrIndex = 0;
+	pDstStr[StrIndex] = ((SrcInt >> 24) & 0xff) - 128;
+	StrIndex++;
+	pDstStr[StrIndex] = ((SrcInt >> 16) & 0xff) - 128;
+	StrIndex++;
+	pDstStr[StrIndex] = ((SrcInt >> 8) & 0xff) - 128;
+	StrIndex++;
+	pDstStr[StrIndex] = (SrcInt & 0xff) - 128;
+	StrIndex++;
+
+	// Ensure null-termination
+	pDstStr[StrIndex] = '\0';
+
+	// Ensure valid UTF-8
+	if(str_utf8_check(pDstStr))
+	{
+		return true;
+	}
+	pDstStr[0] = '\0';
+	return false;
+}
```
This function seems unnecessary and the name doesn't really make it clear what it does unless you know how it's used. The different buffer size requirement also makes it confusing to use together with the other functions. You should be able to use

```cpp
int aInts[2] = {SrcInt, 0x80808000};
char aStr[sizeof(aInts)];
LegacyIntsToStr(aInts, std::size(aInts), aStr, std::size(aStr));
```

to replace the one usage of `LegacyIntToStrPartial`, so this separate function should be unnecessary. It's also currently not tested very much (only on test case), whereas the existing function already has several test cases.

ChillerDragon replied:
This did not compile 

```
/home/chiller/Desktop/git/ddnet/build/src/game/generated/protocol.cpp: In member function ‘int CNetObjHandler::DumpObj(int, const void*, int) const’:
/home/chiller/Desktop/git/ddnet/build/src/game/generated/protocol.cpp:264:28: error: narrowing conversion of ‘2155905024’ from ‘unsigned int’ to ‘int’ [-Wnarrowing]
  264 |         int aInts[2] = {0, 0x80808000};
      |                            ^~~~~~~~~~
```

So I assumed its the order. I tried ``int aInts[2] = {0, 0x00808080};`` that compiled but the strings were cut.

ChillerDragon replied:
for email enjoyers: got it to work. I will delete my confusing comment now :)

ChillerDragon commented:
:white_check_mark: :white_check_mark: :white_check_mark: :rocket:

ChillerDragon commented:
I now also show the names of ddnet ex items

<details>

```diff
diff --git a/datasrc/compile.py b/datasrc/compile.py
index 36f76012a..e9ef66629 100644
--- a/datasrc/compile.py
+++ b/datasrc/compile.py
@@ -254,7 +254,11 @@ void CNetObjHandler::DebugDumpSnapshot(const CSnapshot *pSnap) const
        {
                const CSnapshotItem *pItem = pSnap->GetItem(i);
                int Size = pSnap->GetItemSize(i);
-               dbg_msg("snapshot", "\\t%s type=%d id=%d size=%d", GetObjName(pItem->Type()), pItem->Type(), pItem->Id(), Size);
+               int Type = pSnap->GetItemType(i);
+               const char *pName = GetObjName(pItem->Type());
+               if(Type > OFFSET_UUID)
+                       pName = g_UuidManager.GetName(Type);
+               dbg_msg("snapshot", "\\t%s type=%d id=%d size=%d", pName, pItem->Type(), pItem->Id(), Size);
                if(!DumpObj(pItem->Type(), pItem->Data(), Size))
                        continue;
 
diff --git a/datasrc/seven/compile.py b/datasrc/seven/compile.py
index 790dcc2cd..b09e3ead0 100644
--- a/datasrc/seven/compile.py
+++ b/datasrc/seven/compile.py
@@ -281,7 +281,11 @@ void CNetObjHandler::DebugDumpSnapshot(const ::CSnapshot *pSnap) const
        {
                const CSnapshotItem *pItem = pSnap->GetItem(i);
                int Size = pSnap->GetItemSize(i);
-               dbg_msg("snapshot", "\\t%s type=%d id=%d size=%d", GetObjName(pItem->Type()), pItem->Type(), pItem->Id(), Size);
+               int Type = pSnap->GetItemType(i);
+               const char *pName = GetObjName(pItem->Type());
+               if(Type > OFFSET_UUID)
+                       pName = g_UuidManager.GetName(Type);
+               dbg_msg("snapshot", "\\t%s type=%d id=%d size=%d", pName, pItem->Type(), pItem->Id(), Size);
                if(!DumpObj(pItem->Type(), pItem->Data(), Size))
                        continue;
```

</details>

```
2024-04-12 09:54:20 I snapshot: data_size=356 num_items=10                                                                                                                                                         
2024-04-12 09:54:20 I snapshot:         EX/UUID type=0 id=32767 size=16                                                                                                                                            
2024-04-12 09:54:20 I snapshot:                   0  -1824658838        933dea6a                                                                                                                                   
2024-04-12 09:54:20 I snapshot:                   1   -629591830        da7930ea                                                                                                                                   
2024-04-12 09:54:20 I snapshot:                   2  -1450210576        a98f8af0                                                                                                                                   
2024-04-12 09:54:20 I snapshot:                   3    914991429        3689a945                                                                                                                                   
2024-04-12 09:54:20 I snapshot:         EX/UUID type=0 id=32766 size=16                                                                                                                                            
2024-04-12 09:54:20 I snapshot:                   0    583701389        22ca938d                                                                                                                                   
2024-04-12 09:54:20 I snapshot:                   1    327171627        13803e2b                                                                                                                                   
2024-04-12 09:54:20 I snapshot:                   2  -1636052395        9e7bd255                                                                                                                                   
2024-04-12 09:54:20 I snapshot:                   3  -1901674991        8ea6be11                                                                                                                                   
2024-04-12 09:54:20 I snapshot:         EX/UUID type=0 id=32765 size=16                                                                                                                                            
2024-04-12 09:54:20 I snapshot:                   0   1993229659        76ce455b                                                                                                                                   
2024-04-12 09:54:20 I snapshot:                   1   -102024632        f9eb3a48                                                                                                                                   
2024-04-12 09:54:20 I snapshot:                   2  -1378361269        add7e04b                                                                                                                                   
2024-04-12 09:54:20 I snapshot:                   3  -1810037668        941d045c                                                                                                                                   
2024-04-12 09:54:20 I snapshot:         GameInfo type=6 id=0 size=32                                                                                                                                               
2024-04-12 09:54:20 I snapshot:                   0            0        00000000        m_GameFlags=0 (min=0 max=256)                                                                                              
2024-04-12 09:54:20 I snapshot:                   1            0        00000000        m_GameStateFlags=0 (min=0 max=256)                                                                                         
2024-04-12 09:54:20 I snapshot:                   2            0        00000000        m_RoundStartTick=0 (NetTick)                                                                                               
2024-04-12 09:54:20 I snapshot:                   3            0        00000000        m_WarmupTimer=0 (min=min_int(-2147483648) max=max_int(2147483647))                                                         
2024-04-12 09:54:20 I snapshot:                   4            0        00000000        m_ScoreLimit=0 (min=0 max=max_int(2147483647))                                                                             
2024-04-12 09:54:20 I snapshot:                   5            0        00000000        m_TimeLimit=0 (min=0 max=max_int(2147483647))                                                                              
2024-04-12 09:54:20 I snapshot:                   6            0        00000000        m_RoundNum=0 (min=0 max=max_int(2147483647))                                                                               
2024-04-12 09:54:20 I snapshot:                   7            1        00000001        m_RoundCurrent=1 (min=0 max=max_int(2147483647))                                                                           
2024-04-12 09:54:20 I snapshot:         gameinfo@netobj.ddnet.tw type=32767 id=0 size=12                                                                                                                           
2024-04-12 09:54:20 I snapshot:                   0    247983155        0ec7ec33                                                                                                                                   
2024-04-12 09:54:20 I snapshot:                   1            9        00000009                                                                                                                                   
2024-04-12 09:54:20 I snapshot:                   2           64        00000040   
```

ChillerDragon commented:
Ah lol and now ddnet ex field names too. lol i thought that it would be harder to do.

```
2024-04-12 10:06:35 I snapshot: data_size=356 num_items=10                                                                                                                                                         
2024-04-12 10:06:35 I snapshot:         EX/UUID type=0 id=32767 size=16                                                                                                                                            
2024-04-12 10:06:35 I snapshot:                   0  -1824658838        933dea6a                                                                                                                                   
2024-04-12 10:06:35 I snapshot:                   1   -629591830        da7930ea                                                                                                                                   
2024-04-12 10:06:35 I snapshot:                   2  -1450210576        a98f8af0                                                                                                                                   
2024-04-12 10:06:35 I snapshot:                   3    914991429        3689a945                                                                                                                                   
2024-04-12 10:06:35 I snapshot:         EX/UUID type=0 id=32766 size=16                                                                                                                                            
2024-04-12 10:06:35 I snapshot:                   0    583701389        22ca938d                                                                                                                                   
2024-04-12 10:06:35 I snapshot:                   1    327171627        13803e2b                                                                                                                                   
2024-04-12 10:06:35 I snapshot:                   2  -1636052395        9e7bd255                                                                                                                                   
2024-04-12 10:06:35 I snapshot:                   3  -1901674991        8ea6be11                                                                                                                                   
2024-04-12 10:06:35 I snapshot:         EX/UUID type=0 id=32765 size=16                                                                                                                                            
2024-04-12 10:06:35 I snapshot:                   0   1993229659        76ce455b                                                                                                                                   
2024-04-12 10:06:35 I snapshot:                   1   -102024632        f9eb3a48                                                                                                                                   
2024-04-12 10:06:35 I snapshot:                   2  -1378361269        add7e04b                                                                                                                                   
2024-04-12 10:06:35 I snapshot:                   3  -1810037668        941d045c                                                                                                                                   
2024-04-12 10:06:35 I snapshot:         GameInfo type=6 id=0 size=32                                                                                                                                               
2024-04-12 10:06:35 I snapshot:                   0            0        00000000        m_GameFlags=0 (min=0 max=256)
2024-04-12 10:06:35 I snapshot:                   1            0        00000000        m_GameStateFlags=0 (min=0 max=256)
2024-04-12 10:06:35 I snapshot:                   2            0        00000000        m_RoundStartTick=0 (NetTick)
2024-04-12 10:06:35 I snapshot:                   3            0        00000000        m_WarmupTimer=0 (min=min_int(-2147483648) max=max_int(2147483647))
2024-04-12 10:06:35 I snapshot:                   4            0        00000000        m_ScoreLimit=0 (min=0 max=max_int(2147483647))
2024-04-12 10:06:35 I snapshot:                   5            0        00000000        m_TimeLimit=0 (min=0 max=max_int(2147483647))
2024-04-12 10:06:35 I snapshot:                   6            0        00000000        m_RoundNum=0 (min=0 max=max_int(2147483647))
2024-04-12 10:06:35 I snapshot:                   7            1        00000001        m_RoundCurrent=1 (min=0 max=max_int(2147483647))
2024-04-12 10:06:35 I snapshot:         gameinfo@netobj.ddnet.tw type=32767 id=0 size=12                                                                                                                           
2024-04-12 10:06:35 I snapshot:                   0    247983155        0ec7ec33        m_Flags=247983155                                                                                                          
2024-04-12 10:06:35 I snapshot:                   1            9        00000009        m_Version=9                                                                                                                
2024-04-12 10:06:35 I snapshot:                   2           64        00000040        m_Flags2=64 
```

```diff
diff --git a/datasrc/compile.py b/datasrc/compile.py
index e9ef66629..1624f235c 100644
--- a/datasrc/compile.py
+++ b/datasrc/compile.py
@@ -259,7 +259,7 @@ void CNetObjHandler::DebugDumpSnapshot(const CSnapshot *pSnap) const
                if(Type > OFFSET_UUID)
                        pName = g_UuidManager.GetName(Type);
                dbg_msg("snapshot", "\\t%s type=%d id=%d size=%d", pName, pItem->Type(), pItem->Id(), Size);
-               if(!DumpObj(pItem->Type(), pItem->Data(), Size))
+               if(!DumpObj(Type, pItem->Data(), Size))
                        continue;
 
                for(size_t b = 0; b < Size / sizeof(int32_t); b++)
diff --git a/datasrc/seven/compile.py b/datasrc/seven/compile.py
index b09e3ead0..b01b09fa7 100644
--- a/datasrc/seven/compile.py
+++ b/datasrc/seven/compile.py
@@ -286,7 +286,7 @@ void CNetObjHandler::DebugDumpSnapshot(const ::CSnapshot *pSnap) const
                if(Type > OFFSET_UUID)
                        pName = g_UuidManager.GetName(Type);
                dbg_msg("snapshot", "\\t%s type=%d id=%d size=%d", pName, pItem->Type(), pItem->Id(), Size);
-               if(!DumpObj(pItem->Type(), pItem->Data(), Size))
+               if(!DumpObj(Type, pItem->Data(), Size))
                        continue;
 
                for(size_t b = 0; b < Size / sizeof(int32_t); b++)
```

Robyt3 commented on datasrc/compile.py:
```diff
@@ -242,8 +246,50 @@ def gen_network_source():
+		int Type = pSnap->GetItemType(i);
+		const char *pName = GetObjName(pItem->Type());
+		if(Type > OFFSET_UUID)
+			pName = g_UuidManager.GetName(Type);
```
What happens if there is an item with unknown UUID? Looks like `CUuidManager::GetName` would cause out-of-bounds accesses in this case.

ChillerDragon replied:
Yes good catch! I actually ran into that crash while developing and passing in the wrong type. I totally failed to realize that the debug method should not crash on invalid types! Nice find fixed.

Robyt3 commented on datasrc/seven/compile.py:
```diff
@@ -269,6 +273,43 @@ class CNetObjHandler
+		int Type = pSnap->GetItemType(i);
+		const char *pName = GetObjName(pItem->Type());
+		if(Type > OFFSET_UUID)
+			pName = g_UuidManager.GetName(Type);
```
Same

ChillerDragon commented:
force push

```diff
diff --git a/datasrc/compile.py b/datasrc/compile.py
index 1624f235c..8b7ed3a1a 100644
--- a/datasrc/compile.py
+++ b/datasrc/compile.py
@@ -257,7 +257,12 @@ void CNetObjHandler::DebugDumpSnapshot(const CSnapshot *pSnap) const
                int Type = pSnap->GetItemType(i);
                const char *pName = GetObjName(pItem->Type());
                if(Type > OFFSET_UUID)
-                       pName = g_UuidManager.GetName(Type);
+               {
+                       if(Type < 0 || Type > g_UuidManager.NumUuids())
+                               pName = "unknown";
+                       else
+                               pName = g_UuidManager.GetName(Type);
+               }
                dbg_msg("snapshot", "\\t%s type=%d id=%d size=%d", pName, pItem->Type(), pItem->Id(), Size);
                if(!DumpObj(Type, pItem->Data(), Size))
                        continue;
diff --git a/datasrc/seven/compile.py b/datasrc/seven/compile.py
index b01b09fa7..05f481a6f 100644
--- a/datasrc/seven/compile.py
+++ b/datasrc/seven/compile.py
@@ -284,7 +284,12 @@ void CNetObjHandler::DebugDumpSnapshot(const ::CSnapshot *pSnap) const
                int Type = pSnap->GetItemType(i);
                const char *pName = GetObjName(pItem->Type());
                if(Type > OFFSET_UUID)
-                       pName = g_UuidManager.GetName(Type);
+               {
+                       if(Type < 0 || Type > g_UuidManager.NumUuids())
+                               pName = "unknown";
+                       else
+                               pName = g_UuidManager.GetName(Type);
+               }
                dbg_msg("snapshot", "\\t%s type=%d id=%d size=%d", pName, pItem->Type(), pItem->Id(), Size);
                if(!DumpObj(Type, pItem->Data(), Size))
                        continue;
```

Kaffeine commented on src/engine/shared/legacy_tw_int_strings.h:
```diff
@@ -0,0 +1,11 @@
+
+#include <cstddef>
+
+// Do not use this function unless for legacy code!
```
> For the new code use `%name` instead.

ChillerDragon replied:
What is ``%name`` ?

Kaffeine replied:
I don't know, you tell me. You named a thing 'legacy' and marked it as 'do not use'. Then name a replacement for developers who planned to use this function.

ChillerDragon replied:
No robsti marked it legacy I just moved the code. I also named it Legacy because Heinrich said so as a review in this pr.

This pr is not about marking anything as legacy this pr is about adding details to the snapshot debug method.

I would not have marked it legacy because it is needed for the 0.6 protocol. I assume what the comment means is don't use it for new ddnet objects. But then I also have no idea how new ddnet objects are supposed to add strings to the snap.

ChillerDragon commented:
My bounds check was flawed I fixed it

```diff
diff --git a/datasrc/compile.py b/datasrc/compile.py
index 8b7ed3a1a..a25b71e02 100644
--- a/datasrc/compile.py
+++ b/datasrc/compile.py
@@ -256,13 +256,10 @@ void CNetObjHandler::DebugDumpSnapshot(const CSnapshot *pSnap) const
                int Size = pSnap->GetItemSize(i);
                int Type = pSnap->GetItemType(i);
                const char *pName = GetObjName(pItem->Type());
-               if(Type > OFFSET_UUID)
-               {
-                       if(Type < 0 || Type > g_UuidManager.NumUuids())
-                               pName = "unknown";
-                       else
-                               pName = g_UuidManager.GetName(Type);
-               }
+               if(Type > OFFSET_UUID && Type < g_UuidManager.NumUuids() + OFFSET_UUID)
+                       pName = g_UuidManager.GetName(Type);
+               else
+                       pName = "unknown";
                dbg_msg("snapshot", "\\t%s type=%d id=%d size=%d", pName, pItem->Type(), pItem->Id(), Size);
                if(!DumpObj(Type, pItem->Data(), Size))
                        continue;
diff --git a/datasrc/seven/compile.py b/datasrc/seven/compile.py
index 05f481a6f..51acc3f49 100644
--- a/datasrc/seven/compile.py
+++ b/datasrc/seven/compile.py
@@ -283,13 +283,10 @@ void CNetObjHandler::DebugDumpSnapshot(const ::CSnapshot *pSnap) const
                int Size = pSnap->GetItemSize(i);
                int Type = pSnap->GetItemType(i);
                const char *pName = GetObjName(pItem->Type());
-               if(Type > OFFSET_UUID)
-               {
-                       if(Type < 0 || Type > g_UuidManager.NumUuids())
-                               pName = "unknown";
-                       else
-                               pName = g_UuidManager.GetName(Type);
-               }
+               if(Type > OFFSET_UUID && Type < g_UuidManager.NumUuids() + OFFSET_UUID)
+                       pName = g_UuidManager.GetName(Type);
+               else
+                       pName = "unknown";
                dbg_msg("snapshot", "\\t%s type=%d id=%d size=%d", pName, pItem->Type(), pItem->Id(), Size);
                if(!DumpObj(Type, pItem->Data(), Size))
                        continue;
```

ChillerDragon commented:
Force pushed another fix

```diff
diff --git a/datasrc/compile.py b/datasrc/compile.py
index f1d91f862..b0a75c43b 100644
--- a/datasrc/compile.py
+++ b/datasrc/compile.py
@@ -255,11 +255,9 @@ void CNetObjHandler::DebugDumpSnapshot(const CSnapshot *pSnap) const
                const CSnapshotItem *pItem = pSnap->GetItem(i);
                int Size = pSnap->GetItemSize(i);
                int Type = pSnap->GetItemType(i);
-               const char *pName;
+               const char *pName = GetObjName(pItem->Type());
                if(Type > OFFSET_UUID && Type < g_UuidManager.NumUuids() + OFFSET_UUID)
                        pName = g_UuidManager.GetName(Type);
-               else
-                       pName = "unknown";
                dbg_msg("snapshot", "\\t%s type=%d id=%d size=%d", pName, pItem->Type(), pItem->Id(), Size);
                if(!DumpObj(Type, pItem->Data(), Size))
                        continue;
diff --git a/datasrc/seven/compile.py b/datasrc/seven/compile.py
index fe026bf55..e089e8fba 100644
--- a/datasrc/seven/compile.py
+++ b/datasrc/seven/compile.py
@@ -282,11 +282,9 @@ void CNetObjHandler::DebugDumpSnapshot(const ::CSnapshot *pSnap) const
                const CSnapshotItem *pItem = pSnap->GetItem(i);
                int Size = pSnap->GetItemSize(i);
                int Type = pSnap->GetItemType(i);
-               const char *pName;
+               const char *pName = GetObjName(pItem->Type());
                if(Type > OFFSET_UUID && Type < g_UuidManager.NumUuids() + OFFSET_UUID)
                        pName = g_UuidManager.GetName(Type);
-               else
-                       pName = "unknown";
                dbg_msg("snapshot", "\\t%s type=%d id=%d size=%d", pName, pItem->Type(), pItem->Id(), Size);
                if(!DumpObj(Type, pItem->Data(), Size))
                        continue;
```

Kaffeine commented:
> Force pushed another fix

You don't really have to quote the full diff, as it is always available right there ([Compare](https://github.com/ddnet/ddnet/compare/98caaa4ff97b10b282e6057fdb31b6d30b89414e..3b7a2ad29af01de5574c3406f7630809bef6a72e)).

![image](https://github.com/ddnet/ddnet/assets/374839/f5deb917-9ed1-4dc9-8ca0-ddf5f88b4445)

heinrich5991 commented on datasrc/datatypes.py:
```diff
@@ -383,16 +411,41 @@ def emit_unpack_msg(self):
+		return NetVariable(self.name).emit_dump(offset) + \
+			[f"dbg_msg(\"snapshot\", \"%s\\t{self.name}=%d\", aRawData, pObj->{self.name});"]
+
+class NetTwIntString(NetIntAny):
```
Can you make the constructor of this class take an integer, so that you can generate an array (and make the annoying `m_Skin0` etc. go away, leading to a `m_aSkin[6]` instead)?

ChillerDragon replied:
That is a bit unrelated isn't it? I would consider it out of scope of this pr.

Robyt3 replied:
It's at least somewhat related since you change every associated line in this PR. If it's out of scope for this PR then it could be done first in a separate PR as a refactoring.

ChillerDragon replied:
Oke done.

heinrich5991 replied:
Sorry, I wanted the `NetTwIntString` class to take an integer parameter, not wrap it in a `NetArray`.

Because a string isn't really an array of four-byte segments, but the string is a thing as the whole.

ChillerDragon replied:
I understood you correctly. I just don't like it and intentionally decided against adding the parameter to NetTwIntString. I still made it an `m_aSkin[6]` like you requested (which I still think is out of scope of this pr). We already have an array data type that works might as well use it. No need to define two array data types. And in my head the string is exactly that: an array of four-byte segments.
This is also how the upstream does it.

heinrich5991 replied:
> I understood you correctly. I just don't like it and intentionally decided against adding the parameter to NetTwIntString.

It would be nice to specify that instead of saying "Oke done." then.



> I still made it an `m_aSkin[6]` like you requested (which I still think is out of scope of this pr). We already have an array data type that works might as well use it. No need to define two array data types. And in my head the string is exactly that: an array of four-byte segments.

Having it as a single type e.g. allows us to create methods on it to directly extract the string, or to show the complete string in the debug dump, even if there are multi-byte characters. That's why I liked my solution better.

> This is also how the upstream does it.

Teeworlds isn't upstream to DDNet anymore. It's kinda irrelevant what Teeworlds does.

ChillerDragon replied:
> It would be nice to specify that instead of saying "Oke done." then.

I didn't expect you to care that much.

> Having it as a single type e.g. allows us to create methods on it to directly extract the string, or to show the complete string in the debug dump, even if there are multi-byte characters. That's why I liked my solution better.

The per integer representation seems more useful in the debug dump to me. The multi byte thing could be useful but its also not used so its more a theoretical thing.

> Teeworlds isn't upstream to DDNet anymore. It's kinda irrelevant what Teeworlds does.

Its not irrelavant to me. I still respect teeworlds as the origin for all interesting gameplay, good artwork and engine fundamentals. And I also respect the code decisions that were made there. Also I do work in both code bases and the more they differ the more annoying it is.



All in all im still not really sold on the idea.

heinrich5991 replied:
> I didn't expect you to care that much.

Then just state it, it doesn't hurt in this case.

> Its not irrelavant to me. I still respect teeworlds as the origin for all interesting gameplay, good artwork and engine fundamentals. And I also respect the code decisions that were made there. Also I do work in both code bases and the more they differ the more annoying it is.
> 
> All in all im still not really sold on the idea.

Well, please treat it as irrelevant for DDNet development — we don't want to get held back by a stale codebase.

ChillerDragon replied:
@heinrich5991 what now? Is the string type a requirement for merge or not? I do not like it and I don't see its use case so I am not planning to add it.

ChillerDragon replied:
@heinrich5991 still waiting for your response.

heinrich5991 replied:
> what now? Is the string type a requirement for merge or not?

Yes.

ChillerDragon replied:
Adding new features should not force pr authors to refactor old code. I totally disagree with this maintainer style.

You are forcing me to refactor old code to get in new one. If there really was a problem with the old code there should be a GitHub issue for that and it should be fixed unrelated to this pr.

@heinrich5991 you keep doing this over and over again. I send a pr and you force me to implement something unrelated to the pr which in your opinion is a improvement. Why am I the one who has to do it? Do you really care about these things? Would they be fixed without my pr? You are abusing your leverage that comes with the merge power to force tasks on me that I do not even consider useful.

Why don't you just close the prs when my contributions are not welcome. These annoying random opinions and out of scope requirements are wasting everyone's time.

ChillerDragon replied:
Merged. Thanks for resolving this @Robyt3

heinrich5991 commented on datasrc/datatypes.py:
```diff
@@ -259,6 +259,32 @@ def emit_uncompressed_unpack_and_validate(self, objects):
+		while next_base_name is not None:
+			base_item = only([i for i in objects if i.name == next_base_name])
+			variables = base_item.variables + variables
+			next_base_name = base_item.base
```
This should probably be put into a member function. Maybe `def members_from_this_and_parents()`?

heinrich5991 commented on datasrc/seven/datatypes.py:
```diff
@@ -338,16 +360,34 @@ def emit_unpack(self):
 		self.min = str(min_val)
 		self.max = str(max_val)
+		self.min_val = min_val
+		self.max_val = max_val
```
Saving almost the same info in two member variables sounds annoying. Perhaps you could simply extract the value in `emit_dump` below? E.g. like this:

```py
def fmt_and_arg(val):
    try:
        int(val)
    except ValueError:
        return f"{val}(%d)", f", (int){val}"
    else:
        return val, ""

heinrich5991 reviewed (commented):
Found some more things. Also, the move of the snapshot string functions to engine isn't necessary anymore, is it?

ChillerDragon commented:
Solved conflicts with master.

heinrich5991 commented:
https://github.com/ddnet/ddnet/pull/8189#discussion_r1684282646

ChillerDragon commented:
https://github.com/ddnet/ddnet/pull/8189#discussion_r1788509392

ChillerDragon commented:
Little reminder of how cool this feature is. Here you can see me debugging storing zoom in demos for https://github.com/ddnet/ddnet/issues/8534.
In the middle of the right side you can see the zoom value being updated in real time :rocket:  :rocket:  :rocket:  :rocket:  :rocket: :mag: :rocket:  :rocket:  :rocket:  :rocket: 


https://github.com/user-attachments/assets/2045c2bb-b0df-4094-823e-12e914919523

Robyt3 commented on src/game/client/gameclient.cpp:
```diff
@@ -1636,14 +1636,14 @@ void CGameClient::OnNewSnapshot()
 					CClientData *pClient = &m_aClients[ClientId];
 
-					if(!IntsToStr(&pInfo->m_Name0, 4, pClient->m_aName, std::size(pClient->m_aName)))
+					if(!IntsToStr(pInfo->m_aName, 4, pClient->m_aName, std::size(pClient->m_aName)))
```
There are still more lines where `std::size` could be used for `StrToInts` and `IntsToStr` in `sixup_translate_snapshot.cpp`, `player.cpp` and `demo_extract_chat.cpp`.

ChillerDragon replied:
Oops my bad

Robyt3 commented:
> @Robyt3 something still seems off. It used to be like this where you could properly see how the raw ints built the string

See https://github.com/ChillerDragon/ddnet/pull/24. Seems to work with `CNetObj_ClientInfo` and `CNetObj_SwitchState`.

ChillerDragon commented:
Nice thanks @Robyt3 I fully tested it again and it now works as it should. Ready to merge from my side.

```
2025-07-01 07:56:32 I snapshot:         ClientInfo type=11 id=0 size=68
2025-07-01 07:56:32 I snapshot:                   0  -1008145940        c3e8e9ec        m_aName[0]=-1008145940  IntToStr: Chil
2025-07-01 07:56:32 I snapshot:                   1   -320474428        ece5f2c4        m_aName[1]=-320474428   IntToStr: lerD
2025-07-01 07:56:32 I snapshot:                   2   -220076049        f2e1e7ef        m_aName[2]=-220076049   IntToStr: rago
2025-07-01 07:56:32 I snapshot:                   3   -293568512        ee808000        m_aName[3]=-293568512   IntToStr: n
2025-07-01 07:56:32 I snapshot:                   4   -839520535        cdf5eee9        m_aClan[0]=-839520535   IntToStr: Muni
2025-07-01 07:56:32 I snapshot:                   5   -471302016        e3e88080        m_aClan[1]=-471302016   IntToStr: ch
2025-07-01 07:56:32 I snapshot:                   6  -2139062272        80808000        m_aClan[2]=-2139062272  IntToStr: 
2025-07-01 07:56:32 I snapshot:                   7           64        00000040        m_Country=64
```

edg-l commented:
@ChillerDragon working on this?

ChillerDragon commented:
@edg-l nothing to work on. Ready to merge.

edg-l commented:
> @edg-l nothing to work on. Ready to merge.

can you merge main into this or rebase, ci seems ded

ChillerDragon2 commented:
please rebabe mr bob
